src/compute/src/sink/persist_sink.rs

```diff
@@ -1181,3 +1191,87 @@ where
     let token = Rc::new(shutdown_button.press_on_drop());
     (output_stream, token)
 }
+
+/// This is for REFRESH options on materialized views. It adds an operator that rounds up the
+/// timestamps of data and frontiers to the time of the next refresh. See
+/// `doc/developer/design/20231027_refresh_every_mv.md`.
+///
+/// Note that this currently only works with 1-dim timestamps. (This is not an issue for WMR,
+/// because iteration numbers should disappear by the time the data gets to the Persist sink.)
+fn round_up<G>(
+    coll: Collection<G, Result<Row, DataflowError>, Diff>,
+    refresh_schedule: RefreshSchedule,
+) -> Collection<G, Result<Row, DataflowError>, Diff>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let mut builder = OperatorBuilder::new("round_up".to_string(), coll.scope().clone());
+    let (mut output_buf, output_stream) = builder.new_output();
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
+    builder.build(move |capabilities| {
+        let mut capability = Some(capabilities.into_element());
+        let mut buffer = Vec::new();
+        move |frontiers| {
+            let ac = frontiers.into_element();
+            match ac.frontier().to_owned().into_option() {
+                Some(ts) => {
+                    match refresh_schedule.round_up_timestamp(ts) {
+                        Some(rounded_up_ts) => {
+                            capability.as_mut().unwrap().downgrade(&rounded_up_ts);
+                        }
+                        None => {
+                            // We are past the last refresh. Drop the capability to signal that we
+                            // are done.
+                            capability = None;
+                        }
+                    }
+                }
+                None => {
+                    capability = None;
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
```

Comment:
I think you need to move the capability check into the `input.for_each`. Even if we will never produce output anymore, we still have to drain the operator inputs, otherwise the data that's pushed there from upstream will remain in the input channel and leak.

Reply:
Hmm, good catch!

But actually, I'd like to just terminate the dataflow when we'll never produce output anymore (i.e., when we are past the last ever refresh). I was hoping dropping the capability would already have this effect. Isn't this the case? Or what is the standard way to signal from inside a dataflow that I'm completely done forever?

Reply:
We don't have a way to communicate this from inside the dataflow. Just dropping the capability means the write frontier will advance to the empty frontier, which communicates that the dataflow is not going to write any more data. But upstream operators don't care (or know) about this and continue to do their thing. The `ComputeState` doesn't drop dataflows when they advance to the empty frontier. It only drops them when it receives an `AllowCompaction([])` command from the controller. But the controller doesn't allow compaction to the empty frontier in response to dataflows advancing to the empty frontier, because others might still want to read the produced data.

The current behavior makes sense for indexes, but not necessarily for MVs. I think we could drop MV dataflows immediately when they have advanced to the empty frontier, since we don't need the dataflow to read the data from persist. But we don't have this optimization at the moment because it wasn't useful.

Reply:
+1 to what @teskje said. The operator must not stop consuming its inputs when it doesn't want to produce output anymore, otherwise we'll have a memory (and CPU) leak.

Reply:
Done. Now I've also set the `until` to the last refresh, so we shouldn't ever receive data after we drop our capability when the last refresh is complete.